File: java/src/main/java/net/netheos/pcsapi/credentials/UserCredentialsFileRepository.java
/**
 * Copyright (c) 2014 Netheos (http://www.netheos.net)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package net.netheos.pcsapi.credentials;

import net.netheos.pcsapi.utils.PcsUtils;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.io.Writer;
import java.util.HashMap;
import java.util.Map;

/**
 * This class is a simple reader/writer of users credentials applications information from a plain text file with
 * format:
 * <pre>
 * provider1.app_name_1.user_id1 = json_object provider1.app_name_1.user_id2 = json_object
 * </pre>
 * <p/>
 * Sample code only: data is NOT encrypted in file.
 */
public class UserCredentialsFileRepository
        implements UserCredentialsRepository
{

    private static final Logger LOGGER = LoggerFactory.getLogger( UserCredentialsFileRepository.class );

    private final File file;
    private final Map<String, Credentials> credentialsMap;

    public UserCredentialsFileRepository( File file )
            throws IOException
    {
        this.file = file;

        credentialsMap = new HashMap<String, Credentials>();

        if ( file.exists() ) {
            readFile();
        }
    }

    /**
     * Parses a file of credentials
     *
     * @throws IOException Error reading the file
     */
    private void readFile()
            throws IOException
    {
        BufferedReader reader = null;

        try {
            reader = new BufferedReader( new InputStreamReader( new FileInputStream( file ), PcsUtils.UTF8 ) );

            LOGGER.debug( "Reading credentials file {}", file );

            String line;
            int index = 1;

            while ( ( line = reader.readLine() ) != null ) {
                line = line.trim();

                if ( line.startsWith( "#" ) || line.length() == 0 ) {
                    continue;
                }

                String[] userCredentialsArray = line.split( "=", 2 );

                if ( userCredentialsArray.length != 2 ) {
                    throw new IllegalArgumentException( "Not parsable line #" + index );
                }

                // Key
                final String key = userCredentialsArray[0].trim();
                final Credentials value = Credentials.createFromJson( userCredentialsArray[1].trim() );

                credentialsMap.put( key, value );

                index++;
            }

        } finally {
            PcsUtils.closeQuietly( reader );
        }
    }

    @Override
    public synchronized void save( UserCredentials<?> userCredentials )
            throws IOException
    {
        String key = getUserKey( userCredentials.getAppInfo(), userCredentials.getUserId() );
        credentialsMap.put( key, userCredentials.getCredentials() );
        writeFile();
    }

    @Override
    public synchronized UserCredentials<?> get( AppInfo appInfo, String userId )
    {
        Credentials credentials = null;

        if ( userId != null ) {
            final String key = getUserKey( appInfo, userId );
            credentials = credentialsMap.get( key );

        } else {
            // loop over all credential
            final String prefix = getAppPrefix( appInfo );

            for ( String key : credentialsMap.keySet() ) {
                if ( key.startsWith( prefix ) ) {
                    userId = key.substring( prefix.length() );

                    if ( credentials != null ) {
                        throw new IllegalStateException( "Several user credentials found for application " + appInfo );
                    }

                    credentials = credentialsMap.get( key );
                }
            }
        }

        if ( credentials == null ) {
            throw new IllegalStateException( "No user credentials found for application " + appInfo );
        }

        return new UserCredentials( appInfo, userId, credentials );
    }

    /**
     * Serializes credentials data
     *
     * @throws IOException Error while saving the file
     */
    private void writeFile()
            throws IOException
    {
        LOGGER.debug( "Writing credentials file to {}", file );
        File tempFile = new File( file.getPath() + ".tmp" );

        Writer writer = null;
        try {
            writer = new OutputStreamWriter( new FileOutputStream( tempFile ), PcsUtils.UTF8 );

            for ( String key : credentialsMap.keySet() ) {
                Credentials cred = credentialsMap.get( key );
                final String line = key + "=" + cred.toJson() + "\n";
                writer.write( line );
            }
            writer.flush();

        } finally {
            PcsUtils.closeQuietly( writer );
        }

        file.delete();
        tempFile.renameTo( file );
    }

    /**
     * Builds the key of a credentials according to a given appInfo and userId
     *
     * @param appInfo The application informations
     * @param userId The user identifier
     * @return The user key
     */
    private String getUserKey( AppInfo appInfo, String userId )
    {
        if ( userId == null ) {
            throw new IllegalArgumentException( "userId should not be null" );
        }

        return String.format( "%s%s", getAppPrefix( appInfo ), userId );
    }

    /**
     * Builds a key prefix according to a given appInfo
     *
     * @param appInfo The application informations
     * @return The application prefix
     */
    private String getAppPrefix( AppInfo appInfo )
    {
        /// @TODO scope is ignored in key
        return String.format( "%s.%s.", appInfo.getProviderName(), appInfo.getAppName() );
    }

    @Override
    public String toString()
    {
        return "UserCredentialsRepo{"
               + "file='" + file + '\''
               + ", credentialsMap=" + credentialsMap
               + '}';
    }

}

File: cpp/libpcs_api/src/credentials/user_credentials_file_repository.cc
/**
 * Copyright (c) 2014 Netheos (http://www.netheos.net)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <cstdio>
#include <fstream>  // NOLINT(readability/streams)
#include <string>
#include <memory>
#include <stdexcept>
#include <system_error>
#include <map>

#include "boost/filesystem/operations.hpp"
#include "boost/filesystem/fstream.hpp"
#include "boost/algorithm/string.hpp"
#include "boost/algorithm/string/predicate.hpp"
#include "cpprest/json.h"

#include "pcs_api/user_credentials_file_repository.h"
#include "pcs_api/c_exceptions.h"
#include "pcs_api/internal/logger.h"


namespace pcs_api {


UserCredentialsFileRepository::UserCredentialsFileRepository(
                                        const boost::filesystem::path& path)
    : path_(path) {
    ReadUserCredentialsFile();
}

static std::string GetAppPrefix(const AppInfo& app_info) {
    return utility::conversions::to_utf8string(app_info.provider_name())
            + "."
            + utility::conversions::to_utf8string(app_info.app_name()) + ".";
}

static std::string GetUserKey(const AppInfo& app_info,
                              const std::string& user_id) {
    if (user_id.empty()) {
        BOOST_THROW_EXCEPTION(
            std::invalid_argument("user_id should not be empty"));
    }
    return GetAppPrefix(app_info) + user_id;
}


std::unique_ptr<UserCredentials> UserCredentialsFileRepository::Get(
                                            const AppInfo &app_info,
                                            const std::string& user_id) const {
    std::lock_guard<std::mutex> guard(lock_);

    Credentials *p_credentials = nullptr;
    std::string actual_user_id(user_id);
    if (!user_id.empty()) {
        // Usual case: user_id is defined
        auto key = GetUserKey(app_info, user_id);
        auto it = user_credentials_map_.find(key);
        if (it == user_credentials_map_.end()) {
            BOOST_THROW_EXCEPTION(
                std::invalid_argument(
                    "User " + utility::conversions::to_utf8string(user_id)
                    + " not found for " + app_info.ToString()));
        }
        p_credentials = it->second.get();
    } else {
        // special case where user_id is not defined.
        // Acceptable iff repository contains a single user for this app

        // loop over all credentials:
        // remember that this implementation is not to be used in production
        auto prefix = GetAppPrefix(app_info);
        for (auto it = user_credentials_map_.begin();
             it != user_credentials_map_.end();
             ++it) {
            std::string key = it->first;
            if (boost::starts_with(key, prefix)) {
                // Found:
                actual_user_id = key.substr(prefix.length());
                if (p_credentials != nullptr) {  // already found ?
                    BOOST_THROW_EXCEPTION(
                        std::invalid_argument("Several user credentials found"
                                              " for application " + prefix));
                }
                p_credentials = it->second.get();
                break;
            }
        }
        if (p_credentials == nullptr) {
            BOOST_THROW_EXCEPTION(
                std::invalid_argument("No user credentials found"
                                      " for application " + prefix));
        }
    }
    std::unique_ptr<UserCredentials> p_ret(
                new UserCredentials(app_info, actual_user_id, *p_credentials));
    return p_ret;
}

void UserCredentialsFileRepository::Save(
                                    const UserCredentials& user_credentials) {
    std::lock_guard<std::mutex> guard(lock_);

    auto key = GetUserKey(user_credentials.app_info(),
                          user_credentials.user_id());
    // Update map; we copy given credentials
    std::unique_ptr<Credentials> p_copy(user_credentials.credentials().Clone());
    user_credentials_map_[key] = std::move(p_copy);
    // Serialize to file:
    WriteUserCredentialsFile();
}

UserCredentialsFileRepository::~UserCredentialsFileRepository() {
}

void UserCredentialsFileRepository::ReadUserCredentialsFile() {
    LOG_DEBUG << "Will read UserCredentialsFile: " << path_;
    boost::filesystem::ifstream infofile(path_);
    if (infofile.fail()) {
        std::system_error se(errno, std::system_category());
        const char *p_msg = se.what();
        BOOST_THROW_EXCEPTION(std::ios_base::failure(
                std::string("Could not open file: ")
                + utility::conversions::to_utf8string(path_.c_str())
                + ": " + p_msg));
    }
    for (std::string line; std::getline(infofile, line) ;) {
        // LOG_TRACE << "reading line =" << line;
        boost::algorithm::trim(line);

        if (line.length() == 0 || line[0] == '#') {
            continue;
        }
        // expected line format is key=value:
        std::string::size_type ieq = line.find_first_of('=');
        if (ieq == std::string::npos) {
            LOG_WARN << "Ignored line: not found '='";
            continue;
        }

        std::string key = line.substr(0, ieq);
        boost::algorithm::trim(key);
        std::string credentials_value = line.substr(ieq+1);
        boost::algorithm::trim(credentials_value);
        // LOG_TRACE << "key=" << key;
        // LOG_TRACE << "credentials=" << credentials_value;
        auto p_credentials = std::unique_ptr<Credentials>(
                            Credentials::CreateFromJson(credentials_value));

        user_credentials_map_[key] = std::move(p_credentials);
        LOG_TRACE << "Read credentials for user: " << key;
    }
}

void UserCredentialsFileRepository::WriteUserCredentialsFile() const {
    LOG_DEBUG << "Writing user credentials file to " << path_;
    boost::filesystem::path temp_path = path_;
    temp_path += ".tmp";
    boost::filesystem::ofstream infofile(temp_path);
    if (infofile.fail()) {
        std::system_error se(errno, std::system_category());
        const char *p_msg = se.what();
        BOOST_THROW_EXCEPTION(std::ios_base::failure(
            std::string("Could not open file: ")
                    + utility::conversions::to_utf8string(temp_path.c_str())
                    + ": " + p_msg));
    }

    for (auto it = user_credentials_map_.begin();
         it != user_credentials_map_.end();
         ++it) {
        auto key = it->first;
        auto json = it->second->ToJsonString();
        infofile << key << "=" << json << "\n";
    }
    infofile.flush();
    infofile.close();
    if (!infofile.good()) {
        LOG_ERROR << "Could not write file: " << temp_path;
        BOOST_THROW_EXCEPTION(std::ios_base::failure(
                            "Could not write file: " + temp_path.string()));
    }

    // Rename to final name (need to delete first on windows):
    boost::filesystem::remove(path_);
    boost::filesystem::rename(temp_path, path_);
}


}  // namespace pcs_api
